Validate migration files

realaravinth opened the pull request:
On behalf of [@dachary](https://lab.forgefriends.org/dachary) from the [forgefriends](https://forgefriends.org) project:

JSON Schema validation for data used by Gitea during migrations
Discussion at https://forum.forgefriends.org/t/common-json-schema-for-repository-information/563

---
[source](https://lab.forgefriends.org/forgefriends/forgefriends/-/merge_requests/32)

lunny commented:
I think it's unnecessary to increment the binary size for a test purpose.

realaravinth commented:
> I think it's unnecessary to increment the binary size for a test purpose.

The bindata keyword that was added to .drone.yml and Makefile was removed. Thanks for the quick review :-)

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_763)

realaravinth commented:
on behalf of [@dachary](https://lab.forgefriends.org/dachary):

Thanks for the review. The bug fix was incorrect. The milestone fixtures must have a `deadline_unix` field set instead. I now understand why the year 9999 is special.

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/33#note_794)

realaravinth commented:
The PR was updated to remove the conflicting changes in the vendor directory.

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_818)

Gusted commented on modules/migration/file_format.go:
```diff
@@ -0,0 +1,112 @@
+package migration
+
+import (
+	"encoding/json"
```
Sigh, we really need to get a linter for this specific code.

`encoding/json` should be replaced by `code.gitea.io/gitea/modules/json`

realaravinth replied:
Fixed.

Gusted commented on modules/migration/file_format.go:
```diff
@@ -0,0 +1,112 @@
+func Load(filename string, data interface{}, validation bool) error {
+	isJSON := strings.HasSuffix(filename, ".json")
+
+	_, err := os.Stat(filename)
```
Why is stat necessary? If this is purely to check if the file exists - `ReadFile` should give a error for that as well.

realaravinth replied:
You are correct, removed.

Gusted commented on modules/migration/file_format.go:
```diff
@@ -0,0 +1,112 @@
+		}
+		return err
+	default:
+		return fmt.Errorf("file_format:validate: %T is not a known time", datatype)
```
```suggestion
		return fmt.Errorf("file_format:validate: %T has not a validation implemented", datatype)
```


`Is not a known time` sounds rather confusing to me if I got that error as developer.

realaravinth replied:
Your improved error message was used instead.

Gusted commented on modules/migration/schemas/issue.json:
```diff
@@ -0,0 +1,114 @@
+{
+    "title": "Issue",
+    "description": "Issues associated to a repository within a forge.",
```
Off-topic: what is the definition of `forge` in this specific context?

realaravinth replied:
Examples were added for clarification.

Gusted commented on modules/migration/schemas/issue.json:
```diff
@@ -0,0 +1,114 @@
+
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "$id": "http://example.com/issue.json",
+    "$$target": "issue.json"
```
I had a look around the jsonschema library, but it doesn't seem that `$$target` is used anywhere?

realaravinth replied:
Not yet indeed, but it will be at a later time.

6543 commented:
CI fail related:
```
+ make lint-backend
golangci-lint run --timeout 10m
services/migrations/dump.go:155:3: not enough arguments in call to git.Clone (typecheck)
	})
	 ^
services/migrations/dump.go:162:56: not enough arguments in call to repository.WikiRemoteURL (typecheck)
		wikiRemotePath := repository.WikiRemoteURL(remoteAddr)
		                                                     ^
services/migrations/dump.go:173:5: not enough arguments in call to git.Clone (typecheck)
			}); err != nil {
			 ^
make: *** [Makefile:754: golangci-lint] Error 1
```

realaravinth commented:
@Gusted thanks for your review. I addressed your comments, hopefully to your satisfaction

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_880)

realaravinth commented:
@6543 thanks for highlighting the CI failure. The PR was rebased against today's main branch to resolve them.
---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_881)

Gusted commented on modules/migration/schemas_bindata.go:
```diff
@@ -0,0 +1,10 @@
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+//go:build bindata
```
Could this filename be changed bindata -> static

realaravinth replied:
It is meant to be consistent with the naming of all other bindata files ( BINDATA_DEST := modules/public/bindata.go modules/options/bindata.go modules/templates/bindata.go ) generators ( modules/public/public_bindata.go modules/options/options_bindata.go modules/templates/templates_bindata.go ) and not to be confused with the name of the file that contains the file modules/migration/schemas_static.go.

If that was already clear to you, I may have misunderstood your suggestion?

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_885)

Gusted reviewed (commented):
Only a small nit, otherwise it looks good to me.

lunny commented on modules/migration/milestone.go:
```diff
@@ -9,11 +9,11 @@ import "time"
+	Created     time.Time  `json:"created"`
+	Updated     *time.Time `json:"updated"`
+	Closed      *time.Time `json:"closed"`
+	State       string     `json:"state"`
```
Ah, the comment should be kept.

realaravinth replied:
@lunny good catch! The comment was restored.

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_889)

6543 commented:
please dont force-push I cant easily check what  was changed

6543 reviewed (changes requested):
force push after 2 lgtms - need :eyes:

realaravinth commented:
@6543: Apologies, we'll avoid force pushing next time

6543 commented:
well we squash-merge so its better to not force-push to see the progress
